File: app/src/main/java/com/codylimber/fieldphenology/ui/components/DatasetSelector.kt
package com.codylimber.fieldphenology.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.codylimber.fieldphenology.ui.theme.Primary

data class DatasetItem(
    val key: String,
    val group: String,
    val placeName: String
)

@Composable
fun DatasetSelector(
    datasets: List<DatasetItem>,
    selectedKeys: Set<String>,
    onSelectSingle: (String) -> Unit,
    onToggle: (String) -> Unit,
    onSelectAll: (() -> Unit)? = null
) {
    var expanded by remember { mutableStateOf(false) }

    val displayLabel = when {
        selectedKeys.size == datasets.size && onSelectAll != null -> "All Datasets"
        selectedKeys.size == 1 -> {
            val item = datasets.find { it.key == selectedKeys.first() }
            if (item != null) "${item.group} — ${item.placeName}" else selectedKeys.first()
        }
        else -> "${selectedKeys.size} datasets selected"
    }

    Card(
        modifier = Modifier.fillMaxWidth().clickable { expanded = true },
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 12.dp).fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                displayLabel,
                color = MaterialTheme.colorScheme.onBackground,
                fontSize = 15.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.weight(1f)
            )
            Icon(Icons.Default.ArrowDropDown, "Switch dataset", tint = Primary)
        }
        DropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false },
            modifier = Modifier.fillMaxWidth(0.9f)
        ) {
            if (onSelectAll != null) {
                DropdownMenuItem(
                    text = {
                        Text(
                            "All Datasets",
                            fontWeight = if (selectedKeys.size == datasets.size) FontWeight.Bold else FontWeight.Normal,
                            fontSize = 15.sp,
                            color = Primary
                        )
                    },
                    onClick = { onSelectAll(); expanded = false }
                )
                HorizontalDivider()
            }
            datasets.forEach { item ->
                DropdownMenuItem(
                    text = {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Checkbox(
                                checked = item.key in selectedKeys,
                                onCheckedChange = { onToggle(item.key) },
                                colors = CheckboxDefaults.colors(checkedColor = Primary)
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Column {
                                Text(
                                    item.group,
                                    fontWeight = if (item.key in selectedKeys) FontWeight.Bold else FontWeight.Normal,
                                    fontSize = 15.sp
                                )
                                Text(
                                    item.placeName,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                    fontSize = 12.sp
                                )
                            }
                        }
                    },
                    onClick = { onSelectSingle(item.key); expanded = false }
                )
            }
        }
    }
}

File: ios/Manakin/UI/Components/DatasetSelector.swift
import SwiftUI

struct DatasetItem: Identifiable {
    let key: String
    let group: String
    let placeName: String

    var id: String { key }
}

struct DatasetSelector: View {
    let datasets: [DatasetItem]
    let selectedKeys: Set<String>
    let onSelectSingle: (String) -> Void
    let onToggle: (String) -> Void
    var onSelectAll: (() -> Void)?

    @State private var isExpanded = false
    @Environment(\.appColors) private var colors

    private var displayLabel: String {
        if selectedKeys.count == datasets.count, onSelectAll != nil {
            return "All Datasets"
        } else if selectedKeys.count == 1, let key = selectedKeys.first,
                  let item = datasets.first(where: { $0.key == key }) {
            return "\(item.group) \u{2014} \(item.placeName)"
        } else {
            return "\(selectedKeys.count) datasets selected"
        }
    }

    var body: some View {
        Menu {
            if let onSelectAll {
                Button {
                    onSelectAll()
                } label: {
                    Label("All Datasets", systemImage: selectedKeys.count == datasets.count ? "checkmark.circle.fill" : "circle")
                }

                Divider()
            }

            ForEach(datasets) { item in
                Button {
                    onToggle(item.key)
                } label: {
                    Label {
                        VStack(alignment: .leading) {
                            Text(item.group)
                                .fontWeight(selectedKeys.contains(item.key) ? .bold : .regular)
                            Text(item.placeName)
                                .font(.caption)
                                .foregroundColor(colors.onSurfaceVariant)
                        }
                    } icon: {
                        Image(systemName: selectedKeys.contains(item.key) ? "checkmark.circle.fill" : "circle")
                    }
                }
            }
        } label: {
            HStack {
                Text(displayLabel)
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundColor(colors.onBackground)
                    .lineLimit(1)
                Spacer()
                Image(systemName: "chevron.down")
                    .foregroundColor(.appPrimary)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            .background(colors.surface)
            .clipShape(RoundedRectangle(cornerRadius: 12))
        }
    }
}
